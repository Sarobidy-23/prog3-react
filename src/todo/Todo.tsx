import React, { useState } from 'react';
import './todo.css';
import Item from './Item';
import { todoElement } from '../utils/typeUtils';
import TodoInput from './TodoInput';

function App() {
  const [todo, setTodo] = useState<todoElement[]>([]);

  const switchTodo = (index:number) => {
    let updatedTodo = todo;
    updatedTodo[index].isDone = true
    setTodo([...updatedTodo])
  }

  const insert = (name:string) => {
    let newTodo: todoElement = {name: name, isDone: false};
    setTodo([...todo,newTodo])
  }
 
  return (
    <>
      <div className='flex-container'>
        <TodoInput insert={insert}/>
        <div className='table-container'>
          <div className='todo-container'>
            <label className='todo-label'>TODO</label>
            <section className='todo-element'>
              {todo.length >0 && todo.map((element: todoElement, idx:number)=>(
                (!element.isDone && <Item key={idx} element={element} index={idx} onCheck={switchTodo}/>)
              ))}
            </section>
          </div>
          <div className='done-container'>
            <label className='done-label'>DONE</label>
            <section className='done-element'>
              {todo.length >0 && todo.map((element: todoElement, idx: number)=>(
                    (element.isDone && <Item key={idx} element={element} index={idx} onCheck={switchTodo}/>)
                ))}
            </section>
          </div>
        </div>
      </div>
    </>
  );
}

export default App;
