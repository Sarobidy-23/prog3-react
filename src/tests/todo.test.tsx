/* eslint-disable testing-library/await-async-query */
import renderer, { ReactTestInstance, act }  from 'react-test-renderer';
import Todo from '../todo/Todo';

describe("Global application",()=>{
    const tree = renderer.create(<Todo/>);
    const instance = tree.root as ReactTestInstance;

    it("Snapshot: render todo component", ()=>{
        expect(tree).toMatchSnapshot()
    })
    it("adding new todo and switch to Done", ()=>{

        const input = instance.findByProps({placeholder: "new todo name"});
        
        act(()=>{
            //Enter the todo name
            input.props.onChange({target: {value: "test"}})
        })
        act(()=>{
            //Press "Enter" to add new todo
            input.props.onKeyDown({key:"Enter"})
        })

        const todoContainerInstance = instance.findByProps({className: "todo-element"});
        //The JSX content in Item component
        const todo = todoContainerInstance.findAllByProps({className:"item-container"})
        
        //The item component called in todo component <Item index=0 element={element} />
        const actualTodo = todoContainerInstance.props.children![0] as ReactTestInstance;
        
        //Only the todo name display in the screen
        const todoName = todo[0].children![0]

        const expectedTodo = {'name':'test','isDone':false}
        //New todo is realy added
        expect(JSON.stringify(actualTodo.props.element)).toBe(JSON.stringify(expectedTodo))
        expect(todoName).toBe("test")



        /** Swithed todo to done */
        const checkbox = todo[0].children![1] as ReactTestInstance;
        act(()=>{
            checkbox.props.onChange()
        })

        const doneContainerInstance = instance.findByProps({className: "done-element"});

        //The item component called in todo component <Item index=0 element={element} />
        const actualDone = doneContainerInstance.props.children![0] as ReactTestInstance;
        
        //The JSX content in Item component
        const done = doneContainerInstance.findAllByProps({className:"item-container"})
        //Only the todo name display in the screen
        const doneName = done[0].children![0]

        const expectedDone = {'name':'test','isDone':true}

        //Todo is realy swithed to done
        expect(JSON.stringify(actualDone.props.element)).toBe(JSON.stringify(expectedDone))
        expect(doneName).toBe("test")
       
    })
})
