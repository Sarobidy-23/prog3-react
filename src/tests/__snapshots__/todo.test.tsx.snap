// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global application Snapshot: render todo component 1`] = `
<div
  className="flex-container"
>
  <input
    className="input-new_todo"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="new todo name"
    type="text"
    value=""
  />
  <div>
    <div
      className="todo-container"
    >
      <label
        className="todo-label"
      >
        TODO
      </label>
      <section
        className="todo-element"
      />
    </div>
    <div
      className="done-container"
    >
      <label
        className="done-label"
      >
        DONE
      </label>
      <section
        className="done-element"
      />
    </div>
  </div>
</div>
`;
